### DISCUSSION TITLE
Better logic for DefaultCredentialsProvider

### DISCUSSION THREAD

**Discussion by cfurst** [aws/aws-sdk-java#2873] [General]
It would be nice, since the `DefaultCredentialsProvider` is a big black box, to have the class publish warnings when it cannot actually authenticate completely; even if the elements for authentication are there. Currently the provider just fails silently and moves on in the chain.

For example, and this has cost me much time, let's say `AWS_SHARED_CREDENTIALS_FILE` environment variable is set but the file is not found, instead of just moving on to the next credential provider in the chain, a warning should be printed. Another example, and this is one that really got me, was the `AWS_WEB_IDENTITY_TOKEN_FILE` variable was set, but the sts module was not on the classspath so it didn't use the provider. There are many examples I can think of. A simple warning to let the user know there is probably something mis-configured or unintentionally configured will help developers keep their applications clean and tidy.

Everything helps! As you code, think of the users and not just the algorithm.

Thanks!!

**Comment by yasminetalby** [comment]
Hello @cfurst ,

Thank you very much for your feedback.

This is the repository for the AWS Java SDK v1. Are you referring to AWS Java SDK v2? 

In the AWS Java SDK v2 : `DefaultCredentialsProvider` relies on `AwsCredentialsProviderChain` to  checks sequentially each of places where you can set default credentials and selects the first one you set ([see](https://sdk.amazonaws.com/java/api/latest/software/amazon/awssdk/auth/credentials/DefaultCredentialsProvider.html)).

`AwsCredentialsProviderChain` provides logging at the debug level for each of the failed `provider.resolveCredentials()` ([see](https://github.com/aws/aws-sdk-java-v2/blob/639e57ccc612ea40e95cc8556943fe1fc333d39f/core/auth/src/main/java/software/amazon/awssdk/auth/credentials/AwsCredentialsProviderChain.java#L105)) as well as logging at the debug level of the provider use for successful `provider.resolveCredentials()`([see](https://github.com/aws/aws-sdk-java-v2/blob/639e57ccc612ea40e95cc8556943fe1fc333d39f/core/auth/src/main/java/software/amazon/awssdk/auth/credentials/AwsCredentialsProviderChain.java#L98)).

If you are referring to the AWS Java SDK v1's [DefaultAWSCredentialsProviderChain](https://docs.aws.amazon.com/sdk-for-java/v1/reference/com/amazonaws/auth/DefaultAWSCredentialsProviderChain.html), the behavior is similar:
`DefaultAWSCredentialsProviderChain` extends `AWSCredentialsProviderChain` which also  provides logging at the debug level for each of the failed `provider.getCredentials()` ([see](https://github.com/aws/aws-sdk-java/blob/e660f76952cf40dc0a9473ed70386e498f92d96d/aws-java-sdk-core/src/main/java/com/amazonaws/auth/AWSCredentialsProviderChain.java#L129)) as well as logging at the debug level of the provider use for successful `provider.getCredentials()`([see](https://github.com/aws/aws-sdk-java/blob/e660f76952cf40dc0a9473ed70386e498f92d96d/aws-java-sdk-core/src/main/java/com/amazonaws/auth/AWSCredentialsProviderChain.java#L121)). 

Where you able to retrieve this information? You can find documentation on logging in the [v1](https://docs.aws.amazon.com/sdk-for-java/v1/developer-guide/java-dg-logging.html) and [v2](https://docs.aws.amazon.com/sdk-for-java/latest/developer-guide/logging-slf4j.html) Developer guide.

Is there anything else that could improve your experience? We really appreciate your feedback!

Sincerely,

Yasmine
